Feature/events

jesusrodrz opened the pull request:
1. Added a new generic type to use as the reducer params type, so the se reducer and the dispacth method, can acept a diferent type from `U` and convert it and return it as `T`.
2.  Implement a `subscribeError` and `dispatchError` handle errors cases without create a new event for this cases, this reduce some boilerplate becase most of the situations when we use an event also has to use an errorEvent.
3.  Added a `createAction` function that takes a Event as param, and a async function and binds the event dispatchers, so we can abstract the boilerplate when creating an action, and now only focus on write the actual action behavior.
4. Added the typescript eslint plugin so it not interprect imported and exported types ans unused vars.
5. Implemented a `useXSubscription` hook to avoid conflicts with the previes hook, this name is temporal while we decide the name conventions for this libary

The point 3, is very imporntant because not only reduces boilerplate but allows us now to reuse actions for diferences `purpuse` or `context` in the same view otherwise we can't. Some examples below

before
```js
const onFetchList = createEvent()
const onFetchListError = createEvent()

const fetchList = async (filter) =>{
  let response;
  // validations checks and other logics...

  try {
    response = await ApiFetchResorece(filter)
  } catch (e) {
     onFetchListError.dispatch(e)
     return;
  }
  onFetchList.dispatch(response)
  return response;
}
``` 
now 
```js
const onFetchList = createEvent()

const fetchListAction  = createAction(onFetchList, async (filter) => {
  // validations checks and other logics...
  return =  await ApiFetchResorece(filter)
})

```
Or when we want to used it in diferent contexts...
```js
const fetchListService =  async (filter) => {
  // validations checks and other logics...
  return =  await ApiFetchResorece(filter)
}
const onFetchLatestList = createEvent()
const onFetchFeaturedList = createEvent()

const fetchLatestList  = createAction(onFetchLatesList,fetchListService)
const fetchFeaturedList  = createAction(onFetchLatesList,fetchListService)

// using it
fetchLatestList("latest") // filter set as "latest"
fetchFeaturedList("featured") // filter set as "featured"
// now we can use the fetchList service in the same view without about worry tha the event propagattion overlaps because with this pattern the action are not tied to the events anymore.
```

in a component before
```jsx
const View = () =>{
 useEffect(()=>{
  // we can use the same action so we have to write one for each case
  fetchLatestList("latest")
  fetchFeaturedList("featured")
 },[])
 useSubscription(onFetchLatestList,(list)=>{
 // do something with the list
 })
 useSubscription(onFetchLatestListError,(error)=>{
 // do something with the error
 })
useSubscription(onFetchFeaturedList,(list)=>{
 // do something with the list
 })
 useSubscription(onFetchFeaturedListError,(error)=>{
 // do something with the error
 })
 return <p> component </p>
}
```
now... 
Note that this pattern also have a performance improvements because with the useXSubscription the succes and error cases are subscribed ans unsubscribed in the same useEffect hook that only runs when the components mounts and unmounts
```jsx
const View = () =>{
 useEffect(()=>{
  // here we have diferent referennce but they point to the same action and are bind to their respective event 
  fetchLatestList("latest")
  fetchFeaturedList("featured")
 },[])
 useXSubscription(onFetchLatestList, (list)=>{
 // do something with the list
 }, (error)=>{
 // do something with the error
 })
useXSubscription(onFetchFeaturedList, (list)=>{
 // do something with the list
 }, (error)=>{
 // do something with the error
 })
 return <p> component </p>
}
```

alacret commented:
@jesusrodrz here there is a lot to discuss. For many reasons, including that we don't want to make the library too complex that we enforce some patterns.

`createAction`

I think that we can start with the implementation of the `createAction` decorator, but it has to be very simple, I think that just a function that generates another function and dispatches an event or success and another one on error is simple and powerful enough. 

Let's implement this, and then have a separate discussion to reduce the boilerplate on the success / Error subscription of multiple use cases. 

An idea of this is not to use the word `subscription` because it is misleading. Maybe something like a `useAction` may be a better fit

jesusrodrz commented:
@alacret thats why i preffer to to have that the `createAtion` to handle the pattern, to avoid that the developer has to remenber to create the `onSuccesEvent` and `onErrorEvent`, is much more simple that the event object handle this case by it owns.

Let me explain

Think about a developer who kwons react but not this library, If he need to do a fetch, he needs to create a onSuccessEvent and onError event, then create the actions having mixed the logic that he need for the fetch and the logic of the actions itself (try catch dispacth etc), that what i want to avoid

With the `createAction` and the eventObject that have both cases of success and error is much simpler, only needs to create one Event(that handle both cases) and create the logic he needs and pass to the create action, I don't understand why you think is more complex. What it more complex is the logic of the library itself but it is not necessary a problem, moving complexity to the library and avoid to the end user i think is the best way of aproach the problems

alacret commented:
I think that we all agree on this, I just think that we should do it differently.

`createAction`

- This should take a function, a success event and an error event. So we can avoid the boilerplate of simple actions

```js
const OnProjectCreated = createEvent();
const OnProjectCreatedError = createEvent();

const createProjectAction = createAction((filter) => {
	return await Api.fetchResorece(filter);
}, OnProjectCreated, OnProjectCreatedError);

```

`useAction` (useXSubscription)

- This hook will not have anything to do with events. It will take a function, and register to callbacks, one success callback, and one error callback.

```js
// It doesn't matter if this uses createAction or is just a service function
const createProjectAction = ()=> {

};

const Component = ()=> {
    const localAction = useAction(createProjectAction, (success)=> {}, (error) => {});

	return ()
}

```

jesusrodrz commented:
- The problem i have with this aproach of the `createAction` is that always have  to create an success event and an error event, so why no have it by default in the same event? is simpler, i don't to take a time to understand that i need both. and when using the createAction i dont have to pass two event objects. Explime me why you think is better have the both events because i don't get it

- the use cases of the useXSubscriptions and useActions are complitely diferent, the useXSubscription is to subscribe callback to the event for both cases success and error, this is more perfomance that have two useSubscription separately (this is a fact not an opinion), And i think is more easly to read when having que componen with a large logic because this pattern enforce having the related callbacks for a certain event together, so we dont have the problem of a useSubcription for an event un a line 50 and an useSubcription for the error event in line 150.

- Talking about the useAction like you mentioned earlier this is not much diferent from the usePromise hook that you told you dont wanna use for this cases beacuse it not follow the flux pattern, and in the way you explain only works with services not with action beacuse the actione never throws errors, we talk about throwing error in the actions and cath them with an errorBoundary but in lead volt i notice that some times the aplications breaks anyway so not sure if it is an good idea  

- the way i see a `useAction` hook is this (Be aware that is just and idea and it could be improvable):

```js
const Component = () => {
  // It could be use and event and a service, and under teh hood it uses the `createAction` function 
  const [result,loading,{ error, callMethod }] = useAction(onCreateResource, saveResourceService,(success)=>{ 
   // An success callback to handle the action in and imperative way if i need it, but this if optional 
   // is better to use the declarative way


   },(error)=>{ // An error callback to handle the action in and imperative way if i need it, but this if optional
   })
   
   // Also can be used directly with a action
   const [result,loading,{ error, callMethod }] = useAction(saveResourceAction,(success)=>{ 
   // An success callback to handle the action in and imperative way if i need it, but this if optional 
   // is better to use the declarative way


   },(error)=>{ // An error callback to handle the action in and imperative way if i need it, but this if optional
   })

return <div></div>
}
```

jesusrodrz commented:
@alacret 

I have updated the pr
- Removed the useXSubscription
- the createAction is now like this

``` js
const testAction  = createAction( onSuccessEvent, onErrorEvent, serviceFunction )
```

alacret commented on .eslintrc.js:
```diff
@@ -6,8 +6,14 @@ module.exports = {
+  ignorePatterns: [
+    'lib/**/*.js.',
+    'lib/*.js',
+    'src/actions.ts',
```
why ignore action.ts and store.ts ?

jesusrodrz replied:
Because in those file i use a typescript 4.^ feature and when i did this pr, the eslint version wasn't compatible with the version of typescript used

I have updated the version of eslint parse to a new one that is compatible
